petab/v2/lint.py

```diff
@@ -0,0 +1,549 @@
+
+
+class ValidationResultList(list[ValidationIssue]):
+    """A list of validation results.
+
+    Contains all issues found during the validation of a PEtab problem.
+    """
+
+    def log(
+        self,
+        *,
+        logger: logging.Logger = logger,
+        min_level: ValidationIssueSeverity = ValidationIssueSeverity.INFO,
+    ):
+        """Log the validation results."""
+        for result in self:
+            if result.level < min_level:
+                continue
+            if result.level == ValidationIssueSeverity.INFO:
+                logger.info(result.message)
+            elif result.level == ValidationIssueSeverity.WARNING:
+                logger.warning(result.message)
+            elif result.level >= ValidationIssueSeverity.ERROR:
+                logger.error(result.message)
+
+        if not self:
+            logger.info("PEtab format check completed successfully.")
+
+    def has_errors(self) -> bool:
+        """Check if there are any errors in the validation results."""
+        return any(
+            result.level >= ValidationIssueSeverity.ERROR for result in self
+        )
+
+
+def lint_problem(problem: Problem | str | Path) -> ValidationResultList:
+    """Validate a PEtab problem.
+
+    Arguments:
+        problem: PEtab problem to check. Instance of :class:`Problem` or path
+        to a PEtab problem yaml file.
+    Returns:
+        A list of validation results. Empty if no issues were found.
+    """
+
+    if isinstance(problem, str | Path):
+        problem = Problem.from_yaml(str(problem))
+
+    return problem.validate()
+
+
+class ValidationTask(ABC):
+    """A task to validate a PEtab problem."""
+
+    @abstractmethod
+    def run(self, problem: Problem) -> ValidationIssue | None:
+        """Run the validation task.
+
+        Arguments:
+            problem: PEtab problem to check.
+        Returns:
+            Validation results or ``None``
+        """
+        ...
+
+    def __call__(self, *args, **kwargs):
+        return self.run(*args, **kwargs)
+
+
+class CheckModel(ValidationTask):
+    """A task to validate the model of a PEtab problem."""
+
+    def run(self, problem: Problem) -> ValidationIssue | None:
+        if problem.model is None:
+            return ValidationError("Model is missing.")
+
+        if not problem.model.is_valid():
+            # TODO get actual model validation messages
+            return ValidationError("Model is invalid.")
+
+
+class CheckTableExists(ValidationTask):
+    """A task to check if a table exists in the PEtab problem."""
+
+    def __init__(self, table_name: str):
+        if table_name not in ["measurement", "observable", "parameter"]:
+            # all others are optional
+            raise ValueError(
+                f"Table name {table_name} is not supported. "
+                "Supported table names are 'measurement', 'observable', "
+                "'parameter'."
+            )
```

Comment:
I wouldn't use `ValidationIssueSeverity.*` here, because it's not really a issue with the to-be-validated petab problem, but rather an implementation issue inside the library here.

Actually, it's not caught in  `validate`, because `CheckTableExists` is already instantiated before. `validate` will only execute `.run()`.

Reply:
Oops, I missed that this was in `__init__`. Thanks!